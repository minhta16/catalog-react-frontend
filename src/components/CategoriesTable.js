import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Paper, Grid, MenuItem, MenuList, Typography, Divider,
} from '@material-ui/core';

import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { fetchCategories as fetchCategoriesRedux } from 'references/redux/actions/categories';
import { fetchPosts as fetchPostsRedux } from 'references/redux/actions/posts';

class CategoriesTable extends Component {
  state = {};

  componentDidMount() {
    const { fetchCategories, selectedCatId, fetchPosts } = this.props;
    fetchCategories();
    if (selectedCatId) {
      fetchPosts(selectedCatId);
    }
  }

  render() {
    const {
      categories,
      selectedCatId,
      selectedCatItems,
      fetchPosts,
    } = this.props;
    const selectedCatName = categories[selectedCatId] ? categories[selectedCatId].name : '';
    return (
      <Paper>
        <Grid container spacing={2}>
          <Grid item xs={3} className="vertical-divider">
            <Typography variant="h4" className="mainmenu-categories-header">
              Categories
            </Typography>
            <MenuList>
              {Object.keys(categories).map(key => (
                <MenuItem
                  key={key}
                  selected={selectedCatId === key}
                  component={Link}
                  to={`/${key}`}
                  onClick={() => fetchPosts(key)}
                >
                  {categories[key].name}
                </MenuItem>
              ))}
            </MenuList>
          </Grid>
          <Grid item xs={9}>
            <Typography className="left-margin" variant="h4">
              {selectedCatId ? selectedCatName : 'Choose a category'}
            </Typography>
            <Divider variant="middle" />
            <MenuList>
              {selectedCatId ? (
                Object.keys(selectedCatItems).map(key => (
                  <MenuItem key={key}>{selectedCatItems[key].name}</MenuItem>
                ))
              ) : (
                <Typography variant="body1">
                  Please select a category on the left to view its items.
                </Typography>
              )}
            </MenuList>
          </Grid>
        </Grid>
      </Paper>
    );
  }
}

CategoriesTable.propTypes = {
  selectedCatId: PropTypes.string.isRequired,
  categories: PropTypes.shape.isRequired,
  fetchCategories: PropTypes.func.isRequired,
  selectedCatItems: PropTypes.shape.isRequired,
  fetchPosts: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  selectedCatItems: state.posts,
});

const mapDispatchToProps = dispatch => ({
  fetchCategories: () => dispatch(fetchCategoriesRedux()),
  fetchPosts: id => dispatch(fetchPostsRedux(id)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CategoriesTable);
