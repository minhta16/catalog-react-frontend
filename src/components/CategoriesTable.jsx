import React, { Component } from 'react';
import {
  Paper, Grid, MenuItem, MenuList, Typography, Divider,
} from '@material-ui/core';

class CategoriesTable extends Component {
  state = {
    selectedCategory: { id: -1, name: '' },
    categories: [
      {
        id: 1,
        name: 'Football',
      },
      { id: 2, name: 'Soccer' },
    ],
  };

  componentDidMount() {
    return 0;
  }

  handleItemClick = category => () => {
    this.setState({
      selectedCategory: category,
    });
  };

  render() {
    const { categories, selectedCategory } = this.state;
    return (
      <Paper>
        <Grid container spacing={2}>
          <Grid item xs={3} className="vertical-divider">
            <Typography variant="h4" className="mainmenu-categories-header">
              Categories
            </Typography>
            <MenuList>
              {categories.map(category => (
                <MenuItem
                  key={category.name}
                  selected={selectedCategory.id === category.id}
                  onClick={this.handleItemClick(category)}
                >
                  {category.name}
                </MenuItem>
              ))}
            </MenuList>
          </Grid>
          <Grid item xs={9}>
            <Typography className="left-margin" variant="h4">
              {selectedCategory.name || 'Choose a category'}
            </Typography>
            <Divider variant="middle" />
            <MenuList>
              {categories.map(category => (
                <MenuItem>{category.name}</MenuItem>
              ))}
            </MenuList>
          </Grid>
        </Grid>
      </Paper>
    );
  }
}

export default CategoriesTable;
