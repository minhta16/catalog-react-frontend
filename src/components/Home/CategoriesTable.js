import React, { Component } from 'react';
import PropTypes from 'prop-types';
import randomMC from 'random-material-color';
import {
  Paper,
  Grid,
  MenuList,
  MenuItem,
  ListItem,
  List,
  Typography,
  Divider,
  CircularProgress,
  ListItemText,
  ListItemAvatar,
  Avatar,
} from '@material-ui/core';

import { Link } from 'react-router-dom';

class CategoriesTable extends Component {
  state = {};

  getShortDescription = (description) => {
    let retString = description.substr(0, 100);
    retString += description.length > 100 ? '...' : '';
    return retString;
  };

  getSecondaryText = (item) => (
    <React.Fragment>
      <Typography component="span" variant="body2" style={{ display: 'inline' }}>
        {`$${Number(item.price).toFixed(2)} - `}
      </Typography>
      {this.getShortDescription(item.description)}
    </React.Fragment>
  );

  render() {
    const {
      categories,
      selectedCatId,
      selectedCatItems,
      fetchPosts,
      selectedCat,
      categoriesLoading,
      postsLoading,
    } = this.props;
    const selectedCatName = selectedCat ? selectedCat.name : '';
    return (
      <Paper>
        {/* Create a loading animation if the categories are loading */}
        {categoriesLoading ? (
          <CircularProgress />
        ) : (
          <Grid container spacing={2}>
            <Grid item xs={3} className="vertical-divider">
              <Grid className="mainmenu-categories-header">
                <Typography variant="h4">Categories</Typography>
              </Grid>
              <MenuList>
                {categories.map((category) => (
                  <MenuItem
                    className="categoryMenuItem"
                    key={category.id}
                    selected={String(selectedCatId) === String(category.id)}
                    component={Link}
                    to={`/category/${category.id}`}
                    onClick={() => fetchPosts(category.id)}
                  >
                    {category.name}
                  </MenuItem>
                ))}
              </MenuList>
            </Grid>
            <Grid item xs={9}>
              {/* Create a loading animation if the posts are loading */}
              {postsLoading ? (
                <CircularProgress />
              ) : (
                <div>
                  <Typography id="categories-table-name-typo" className="left-margin" variant="h4">
                    {selectedCatId ? selectedCatName : 'Choose a Category'}
                  </Typography>
                  <Divider variant="middle" />
                  {/* A list of items to be selected, click to redirect */}
                  <List>
                    {/* If selectedCatID is available then display the menu. If not, the right side will display a message telling the user to select a category */}
                    {selectedCatId ? (
                      selectedCatItems.map((item) => (
                        <ListItem
                          button
                          className="itemsMenuItem"
                          key={item.id}
                          component={Link}
                          to={`/category/${selectedCatId}/post/${item.id}`}
                        >
                          <ListItemAvatar>
                            <Avatar
                              style={{
                                backgroundColor: randomMC.getColor({
                                  shades: ['300'],
                                  text: item.name,
                                }),
                                color: '#fff',
                              }}
                            >
                              {item.name.charAt(0)}
                            </Avatar>
                          </ListItemAvatar>
                          <ListItemText
                            primary={item.name}
                            secondary={this.getSecondaryText(item)}
                          />
                        </ListItem>
                      ))
                    ) : (
                      <Typography className="itemTypography" variant="body1">
                        Please select a category on the left to view its items.
                      </Typography>
                    )}
                  </List>
                </div>
              )}
            </Grid>
          </Grid>
        )}
      </Paper>
    );
  }
}

CategoriesTable.propTypes = {
  selectedCatId: PropTypes.string,
  categories: PropTypes.array.isRequired,
  selectedCatItems: PropTypes.array.isRequired,
  fetchPosts: PropTypes.func.isRequired,
  selectedCat: PropTypes.object.isRequired,
  categoriesLoading: PropTypes.bool.isRequired,
  postsLoading: PropTypes.bool.isRequired,
};

CategoriesTable.defaultProps = {
  selectedCatId: '',
};

export default CategoriesTable;
