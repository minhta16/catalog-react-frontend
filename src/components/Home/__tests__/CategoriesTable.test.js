/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import CategoriesTable from 'components/Home/CategoriesTable';

describe('components/CategoriesTable', () => {
  let wrapper;
  let props;
  // let warning;

  // const update = () => {
  //   wrapper.update();
  //   warning = wrapper.find(Warning);
  // };

  const setup = () => {
    wrapper = shallow(<CategoriesTable {...props} />);
  };

  beforeEach(() => {
    props = {
      selectedCatId: '1',
      onClick: jest.fn(),
      categories: [
        { created: '2019', updated: '2019', description: 'testDesc', name: 'testName', id: '1' },
      ],
      selectedCat: {
        created: '2019',
        updated: '2019',
        description: 'testDesc',
        name: 'testName',
        id: '1',
      },
      selectedCatItems: [],
      fetchPosts: jest.fn(),
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should display no names if selectedCat is not present', () => {
    props.selectedCat = {};
    setup();
    expect(wrapper.find('#categories-table-name-typo').text()).toBe('');
  });

  it('should display a name  if selectedCat is  present', () => {
    setup();
    expect(wrapper.find('#categories-table-name-typo').text()).toBe(props.selectedCat.name);
  });

  it('should call fetchPosts when a category is clicked', () => {
    setup();
    const { fetchPosts } = props;
    wrapper.findWhere((node) => node.key() === '1').simulate('click');
    expect(fetchPosts).toBeCalled();
  });

  it('should display items if selectedCatId is provided', () => {
    props.selectedCatItems = [
      {
        name: 'hihi',
      },
      {
        name: 'hihi',
      },
    ];
    props.selectedCatId = '1';
    setup();
    expect(wrapper.find('.itemsMenuItem').length).toBe(2);
  });

  it('should display a Typography if selectedCatId is not provided', () => {
    props.selectedCatItems = [
      {
        name: 'hihi',
      },
      {
        name: 'hihi',
      },
    ];
    props.selectedCatId = '';
    setup();
    expect(wrapper.find('.itemTypography').length).toBe(1);
  });
});
