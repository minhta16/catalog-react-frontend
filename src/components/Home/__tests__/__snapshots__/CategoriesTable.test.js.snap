// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CategoriesTable should render correctly 1`] = `
<WithStyles(ForwardRef(Paper))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    spacing={2}
  >
    <WithStyles(ForwardRef(Grid))
      className="vertical-divider"
      item={true}
      xs={3}
    >
      <WithStyles(ForwardRef(Grid))
        className="mainmenu-categories-header"
      >
        <WithStyles(ForwardRef(Typography))
          variant="h4"
        >
          Categories
        </WithStyles(ForwardRef(Typography))>
      </WithStyles(ForwardRef(Grid))>
      <ForwardRef(MenuList)>
        <WithStyles(ForwardRef(MenuItem))
          className="categoryMenuItem"
          component={[Function]}
          key="1"
          onClick={[Function]}
          selected={true}
          to="/category/1"
        >
          testName
        </WithStyles(ForwardRef(MenuItem))>
      </ForwardRef(MenuList)>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      item={true}
      xs={9}
    >
      <div>
        <WithStyles(ForwardRef(Typography))
          className="left-margin"
          id="categories-table-name-typo"
          variant="h4"
        >
          testName
        </WithStyles(ForwardRef(Typography))>
        <WithStyles(ForwardRef(Divider))
          variant="middle"
        />
        <WithStyles(ForwardRef(List)) />
      </div>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Paper))>
`;
